### DISCUSSION TITLE
Volatile Read Implications

### DISCUSSION THREAD

**Discussion by wsugarman** [dotnet/runtime#76282] [Q&A]
I was recently reading [Lazy.cs](https://github.com/dotnet/runtime/blob/v6.0.9/src/libraries/System.Private.CoreLib/src/System/Lazy.cs) and noticed that the member [`_state` is `volatile`](https://github.com/dotnet/runtime/blob/163a63591cf9e9b682063cf3995948c2b885a042/src/libraries/System.Private.CoreLib/src/System/Lazy.cs#L191). I _think_ this is done to ensure proper _acquire semantics_ and _release semantics_ (as per the [language specification](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/classes#1454-volatile-fields)) for the [`Value`](https://github.com/dotnet/runtime/blob/163a63591cf9e9b682063cf3995948c2b885a042/src/libraries/System.Private.CoreLib/src/System/Lazy.cs#L505) property. More specifically, if `_state` is `null`, it must mean that `_value` is non-`null`, because `_state` is assigned last and none of the other writes like `_value` can be reordered after it (there are also helpful comments to emphasize this point). 

However, does the `volatile` keyword also imply any behavior about the "freshness" of the other members access by a given thread? Is it possible that an old value of the non-`volatile` member `_value` is cached in some register by the time we read it? Perhaps this is processor-specific?

I see that there are really three code paths for `Lazy<T>` values, so are there "freshness" guarantees made through the use of a `lock` for `LazyThreadSafetyMode.ExecutionAndPublication` and `Interlocked` for `LazyThreadSafetyMode.PublicationOnly` that are not present for `LazyThreadSafetyMode.None`?

**Comment by huoyaoyuan** [comment]
The CLR memory model documentation (#75790) may provide some help for you.
